flash/image/classification/adapters.py

```diff
@@ -0,0 +1,416 @@
+        self.head = head
+        self.algorithm_cls = algorithm_cls
+        self.ways = ways
+        self.kshots = kshots
+        self.queries = queries
+
+        params = inspect.signature(self.algorithm_cls).parameters
+
+        algorithm_kwargs["train_ways"] = ways
+        algorithm_kwargs["test_ways"] = ways
+
+        algorithm_kwargs["train_shots"] = kshots - queries
+        algorithm_kwargs["test_shots"] = kshots - queries
+
+        algorithm_kwargs["train_queries"] = queries
+        algorithm_kwargs["test_queries"] = queries
+
+        if "model" in params:
+            algorithm_kwargs["model"] = Model(backbone=backbone, head=head)
+
+        if "features" in params:
+            algorithm_kwargs["features"] = Model(backbone=backbone, head=None)
+
+        if "classifier" in params:
+            algorithm_kwargs["classifier"] = head
+
+        self.model = self.algorithm_cls(**algorithm_kwargs)
+
+        # this algorithm requires a special treatment
+        self._algorithm_has_validated = self.algorithm_cls != l2l.algorithms.LightningPrototypicalNetworks
+
+    def _default_transform(self, dataset) -> List[Callable]:
+        return [
+            l2l.data.transforms.FusedNWaysKShots(dataset, n=self.ways, k=self.kshots),
+            l2l.data.transforms.LoadData(dataset),
+            RemapLabels(dataset),
+            l2l.data.transforms.ConsecutiveLabels(dataset),
+        ]
+
+    @property
+    def task(self) -> Task:
+        return self._task.task
+
+    def convert_dataset(self, dataset):
+        metadata = getattr(dataset, "data", None)
+        if metadata is None or (metadata is not None and not isinstance(dataset.data, list)):
+            raise MisconfigurationException("Only dataset built out of metadata is supported.")
+
+        indices_to_labels = {index: sample[DefaultDataKeys.TARGET] for index, sample in enumerate(dataset.data)}
+        labels_to_indices = defaultdict(list)
+        for idx, label in indices_to_labels.items():
+            labels_to_indices[label].append(idx)
+
+        if len(labels_to_indices) < self.ways:
+            raise MisconfigurationException(
+                "Provided `ways` should be lower or equal to number of classes within your dataset."
+            )
+
+        if min(len(indice) for indice in labels_to_indices.values()) < (self.kshots + self.queries):
+            raise MisconfigurationException(
+                "Provided `kshots` should be lower than the lowest number of sample per class."
+            )
+
+        # convert the dataset to MetaDataset
+        dataset = l2l.data.MetaDataset(
+            dataset, indices_to_labels=indices_to_labels, labels_to_indices=labels_to_indices
+        )
+        taskset = l2l.data.TaskDataset(
+            dataset=dataset,
+            task_transforms=self._default_transform(dataset),
+            num_tasks=-1,
+            task_collate=self._identity_fn,
+        )
+        dataset = Epochifier(taskset, 100)
```

Comment:
### `num_epochs`, `num_task`, `meta_batch_size`, `kshots` and `batch_size`
Here, it seems a bit arbitrary to force the number of tasks per epoch to be 100. We should aim to replicate the training loop define, for example, [here](https://github.com/learnables/learn2learn/blob/06893e847693a0227d5f35a6e065e6161bb08201/examples/vision/maml_omniglot.py#L84-L129). In other words, ideally, the user should be able to choose the
- [x] number of epochs -> done in the trainer
- [x] how many task to create in total -> done by specifying the `num_task` in the l2l backend, ideally to be exposed to user
- [ ] number of tasks per epoch aka `meta_batch_size` -> ideally, to be exposed to user and can default to `num_task`
- [x] how many observations per each task -> controlled by `kshots` and dealt with by l2l in the backend
- [ ] the batch_size within the tasks -> suppose that `kshots` for each task is big enough that data do not fit in memory. Note that this can potentially avoid the cumbersome fact that in the datamodule the user is forced to set `batch_size` to 1 otherwise `process_{train, val, test}_dataset` produces an AssertionError

Reply:
Done ! Thanks.